### PR TITLE
terminate loading animation for 404 imgaes

### PR DESCRIPTION
**Pull request by zhouyx**
#3806

### CODE REVIEW

**Comment by zhouyx** [comment]
Added a `loadingError_` flag and terminates the animation. Is this on the right track?

File: src/custom-element.js [hunk 1]
```diff
@@ -797,6 +800,7 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
         this.implementation_.firstLayoutCompleted();
       }
     }, reason => {
+      this.loadingError_ = true;
       this.toggleLoading_(false, /* cleanup */ true);
```
Comment (cramforce) [inline src/custom-element.js]: Why wasn't this effective?

File: src/custom-element.js [hunk 2]
```diff
@@ -381,6 +381,9 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
      * @private {boolean|undefined}
      */
     this.isInTemplate_ = undefined;
+
+    /** @private {boolean} */
+    this.loadingError_ = false;
```
Comment (cramforce) [inline src/custom-element.js]: maybe `hadLoadingError_`?

File: src/custom-element.js [hunk 3]
```diff
@@ -1118,6 +1122,9 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
    * @private @final @this {!Element}
    */
   ElementProto.toggleLoading_ = function(state, opt_cleanup) {
+    if (this.loadingError_) {
```
Comment (cramforce) [inline src/custom-element.js]: Don't fully understand this. In any case it needs a big comment.

**Comment by cramforce** [comment]
Could you explain in the description what this is changing and why it works.

Tests are needed.

**Comment by erwinmombay** [comment]
does the state reset on resize? (where we fetch a possibly new resource)

**Comment by zhouyx** [comment]
In `viewportCallback()` there's a time delay of 100 before call function `this.toggleLoading_(true)`. This creates a race condition of the `this.toggleLoading_(true)` call in `viewportCallback()` and the `this.toggleLoading_(false, /* cleanup */ true);` call in `layoutCallback()`

File: src/custom-element.js (hunk 1 above)
Comment (zhouyx) [inline src/custom-element.js]: `this.toggleLoading_(true)` in `viewportCallback` could be called later than this

**Comment by zhouyx** [comment]
@erwinmombay I think we definitely need to reset the state when fetching a new resource

**Comment by jridgewell** [comment]
> In viewportCallback() there's a time delay of 100 before call function this.toggleLoading_(true)

Yes, but I don't think this is the right way to accomplish this. Part of `#toggleLoading` is to check if `#isLoadingEnabled` is `true`. We should make that return `false` if we have a load error, instead of adding new checks.

- - -

@dvoytenko: Why do we go back and forth from loading state? Shouldn't it just go from the loading state -> laid out (or attempted) state?

**Comment by zhouyx** [comment]
@jridgewell sure I will add another check in `isLoadingEnabled` and minimize changes elsewhere. 
@erwinmombay I reset `hadLoadingError_` flag in `layoutCallback()` in case it's been called multiple times. @cramforce Is `layoutCallback()` the only place where an element will try to load its resources? 
test added, PTAL

File: src/custom-element.js [hunk 4]
```diff
@@ -1076,14 +1086,16 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
...
+    // 7. The element has loading error.
     if (this.loadingDisabled_ === undefined) {
       this.loadingDisabled_ = this.hasAttribute('noloading');
     }
     if (this.loadingDisabled_ || !isLoadingAllowed(this.tagName) ||
         this.layoutWidth_ < MIN_WIDTH_FOR_LOADING_ ||
         this.layoutCount_ > 0 ||
-        isInternalOrServiceNode(this) || !isLayoutSizeDefined(this.layout_)) {
+        isInternalOrServiceNode(this) || !isLayoutSizeDefined(this.layout_) ||
+        this.hadLoadingError_) {
```
Comment (zhouyx) [inline src/custom-element.js]: Put the `hadLoadingError_` check inside `isLoadingEnabled_` function, since we already check `isLoadingEnabled` before add loading dots. 
Also it's reasonable to disable loading when having error.

File: test/manual/amp-img-404.html [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<!doctype html>
```
Comment (zhouyx) [inline test/manual/amp-img-404.html]: Do I need to remove this file later?

File: src/custom-element.js [hunk 6]
```diff
@@ -549,7 +552,11 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
         // Few top elements will also be pre-initialized with a loading
         // element.
         this.getVsync_().mutate(() => {
-          this.prepareLoading_();
+          // Repeat "loading enabled" check because it could have changed while
+          // waiting for vsync.
+          if (this.isLoadingEnabled_()) {
+            this.prepareLoading_();
+          }
```
Comment (zhouyx) [inline src/custom-element.js]: `prepareLoading_()` 's caller must ensure that loading can actually be shown. I add another check in case `isLoadingEnabled_()` value changes during the wait for vsync

**Comment by zhouyx** [comment]
I put the `hadLoadingError_` check inside `isLoadingEnabled_` function. And `isLoadingEnabled_` is currently being checked inside `toggleLoading(true).

ready for review, PTAL

File: test/manual/amp-img-404.html (hunk 5 above)
Comment (cramforce) [inline test/manual/amp-img-404.html]: Nope.

**Comment by cramforce** [comment]
This code is really hard to follow.

@jridgewell want to take another look?

While this doesn't really look good to me, I think it makes sense :)

File: src/custom-element.js [hunk 7]
```diff
@@ -381,6 +381,9 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
      * @private {boolean|undefined}
      */
     this.isInTemplate_ = undefined;
+
+    /** @private {boolean} */
+    this.hadLoadingError_ = false;
```
Comment (jridgewell) [inline src/custom-element.js]: I think this is definitely the right way to solve it, but we can make it a little smarter by using a counter (like `#layoutCount`) instead of a boolean. We can then check if `#layoutAttempts` (naming?) is `> 0` and refuse to show loading.

File: src/custom-element.js [hunk 8]
```diff
@@ -785,6 +792,8 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
     this.assertNotTemplate_();
     dev.assert(this.isUpgraded() && this.isBuilt(),
         'Must be upgraded and built to receive viewport events');
+    // Reset hadLoadingError_ when trying to load resources.
+    this.hadLoadingError_ = false;
```
Comment (jridgewell) [inline src/custom-element.js]: I don't think this is the right location. `#layoutCallback` should only be called once (whether is succeeds or fails), until `#unlayoutCallback` is called. I think setting this in `#unlayoutCallback` with the `layoutCount = 0` is the right spot.

File: src/custom-element.js (hunk 7 above)
Comment (zhouyx) [inline src/custom-element.js]: By using a counter, when will `#layoutAttempts` be greater than if `layoutCallback` is only called once?

Comment (jridgewell) [inline src/custom-element.js]: Good point. Looking into it more, an element can get have `#layoutCallback` multiple times if it defines a `#isRelayoutNeeded`. So ignore that part. 😬

Comment (zhouyx) [inline src/custom-element.js]: I'll use the existing `this.layoutCount_` value, I agree it's definitely better, and terminate loading animation for element should be the same process despite loading status.

**Comment by zhouyx** [comment]
When trying to increase `this.layoutCount_` in `layoutCallback` for loading error, I found we don't really care what the `this.layoutCount_` value is, it's either 0, 1 or some value greater than 1. 

Currently `this.layoutCount_ == 1` is the condition to call `firstLayoutCompleted()` after successfully laid out. If I want to update `layoutCount_` when having load error, I'll need another variable for this `firstLayoutCompleted` condition then. (probably sth like `this.isFirstLayout_ = true`);

Another thing is because we don't need the value for `this.layoutCount_`, we could even change it to a boolean value like `this.isLayoutCalled_`. @jridgewell what do you think about this?

File: src/custom-element.js [hunk 9]
```diff
@@ -793,10 +802,15 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
...
-      if (this.layoutCount_ == 1) {
+      /* what to do here, if load fail, but layout for the second time. */
+      /* TODO: zhouyx, still need a way to figure out how to call firstLayoutCompleted() */
+      if (this.isFirstLayout_) {
+        this.isFirstLayout_ = false;
         this.implementation_.firstLayoutCompleted();
       }
     }, reason => {
+      this.layoutCount_++;
+      this.hadLoadingError_ = true;
```
Comment (zhouyx) [inline src/custom-element.js]: needs to be deleted

File: src/custom-element.js [hunk 10]
```diff
@@ -793,10 +800,15 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
       this.readyState = 'complete';
       this.layoutCount_++;
       this.toggleLoading_(false, /* cleanup */ true);
-      if (this.layoutCount_ == 1) {
```
Comment (jridgewell) [inline src/custom-element.js]: Why is this changed?

Comment (zhouyx) [inline src/custom-element.js]: Here's what I understand, say if `layoutCallback` has loading error first time and success the second time. `this.layoutCount_` will equal to 2 here, and `firstLayoutCompleted()` will never be called.

Comment (jridgewell) [inline src/custom-element.js]: 👍. Without @dvoytenko to rule here, I think I'd rather keep track of two separate ints for now. We can create a separate issue to merge them once he decides.

File: src/custom-element.js [hunk 11]
```diff
@@ -792,11 +799,16 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
...
       this.layoutCount_++;
+      this.layoutAttemptCount_++;
       this.toggleLoading_(false, /* cleanup */ true);
+      // Check if this is the first success layout that needs to call
+      // to call firstLayoutCompleted.
       if (this.layoutCount_ == 1) {
         this.implementation_.firstLayoutCompleted();
       }
     }, reason => {
+      // add layoutCount_ by 1 despite load fails or not
```
Comment (zhouyx) [inline src/custom-element.js]: add `layoutAttemptCount_` by 1

**Comment by jridgewell** [comment]
LGTM.

**Comment by zhouyx** [comment]
Fix tests, by changing `layoutCount_` to `layoutAttemptCount_`

**Comment by dvoytenko** [comment]
@jridgewell no need to go back and forth. we only care about the first layout generically. Components that need more can call toggle API directly. However, image resize should not be triggering loading animation.

**Comment by zhouyx** [comment]
What's your idea of merging the two variables `layoutCount_` and `layoutAttempCount_` (I added in this PR), and introducing another `isFirstLayout` boolean variable to indicate the first layout? @dvoytenko 
If yes, should I have another PR for the change and merge this one first?

**Comment by dvoytenko** [comment]
@zhouyx Responded separately on a bug. But, I think it's fine to do this. If you decide to do this - let's do this in this PR.

**Comment by zhouyx** [comment]
@dvoytenko PTAL

File: src/custom-element.js [hunk 12]
```diff
@@ -329,6 +329,9 @@ export function createAmpElementProto(win, name, opt_implementationClass) {
     this.layoutCount_ = 0;
 
     /** @private {boolean} */
+    this.isFirstLayoutCompleted = true;
```
Comment (dvoytenko) [inline src/custom-element.js]: `@private` name should be terminated with underscore.

**Comment by dvoytenko** [comment]
LGTM with one nit. The naming is important since we drive var renaming based on this. Please fix and feel free to merge.
